### ISSUE TITLE
Exception is not being raised after adding exception to 'throws' in the IDL

### ISSUE THREAD

**Issue by collenjones** [uber/tchannel-python#300]
I am having a strange error where the below test fails with InvalidArgument in the throws section in the IDL below, BUT it passes if I remove it. I have confirmed that the exception is being thrown when I step through the code.

Any ideas why this exception would be swallowed?

Thrift version: 0.9.3
I am loading the thrift spec with thrift.load()

===IDL===
```
service DataAggregationService 
{
  ...
  void tripCreatedOrChanged(
    1: TripCreatedOrChangedRequest request
  ) throws (
    1: ServiceError serviceError,
    2: InvalidArgument invalidArgument
  )
}
```

===Handler===
```
@stats.wrapper("servicename.handlers.tripCreatedOrChanged")
@tchannel.thrift.register(servicename.DataAggregationService)
@thrift_handler_wrapper
@coroutine
def tripCreatedOrChanged(request):
    trip_uuid = validate_uuid(request.body.tripUuid)
    yield DataAggregationService.get_and_log_profile_trip_changes(trip_uuid)
```
NOTE: `validate_uuid` raises an InvalidArgument exception if the uuid is not a valid UUID.

===Test===
```
 @mock.patch.object(DataAggregationService, 'get_and_log_profile_trip_changes')
    @gen_test
    def test_trip_created_or_changed_handles_invalid_uuid(self, mock_get_log):
        """Ensure tripCreatedOrChanged raises an InvalidArgument exception if uuid is invalid."""
        request = MockRequest(tripUuid='invalid-uuid')
        with self.assertRaises(InvalidArgument):
            yield data_aggregation.tripCreatedOrChanged(request)

```

Stacktrace:
```
_______________________________________________ TestDataAggregationHandler.test_user_profile_created_or_changed_handles_invalid_uuid _______________________________________________
env/local/lib/python2.7/site-packages/tornado/testing.py:525: in post_coroutine
    timeout=timeout)
env/local/lib/python2.7/site-packages/tornado/ioloop.py:453: in run_sync
    return future_cell[0].result()
env/local/lib/python2.7/site-packages/tornado/concurrent.py:232: in result
    raise_exc_info(self._exc_info)
env/local/lib/python2.7/site-packages/tornado/gen.py:1017: in run
    yielded = self.gen.send(value)
tests/handlers/test_data_aggregation.py:49: in test_user_profile_created_or_changed_handles_invalid_uuid
    yield data_aggregation.userProfileCreatedOrChanged(request)
E   AssertionError: InvalidArgument not raised
```

**Comment by abhinav** [comment]
What's `thrift_handler_wrapper`? Could it be catching the exception?

**Comment by collenjones** [comment]
Thanks for the stats.wrapper tip!

`thrift_handler_wrapper` is catching the exception, but also reraising it.

```
def thrift_handler_wrapper(func):
    """Wraps the handler methods with exceptions."""
    @coroutine
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        try:
            result = yield maybe_future(func(*args, **kwargs))

        except (exceptions.ServiceAError,
                exceptions.ServiceBError) as e:
            raise servicename.ServiceError(message='External service exception: {}'.format(str(e)))
        except servicename.InvalidArgument as e:
            raise e
        except Exception as e:
            raise servicename.ServiceError(message='Uncaught exception: {}'.format(str(e)))

        raise Return(result)
    return wrapper
```

**Comment by abhinav** [comment]
Are you using `tchannel.thrift.load` or using code generated by Apache Thrift? You can't mix them. If you're raising the `InvalidArgument` generated by `tchannel.thrift.load`, you catch clause can't use the one generated by Apache Thrift.

You can verify this by stepping through the code and comparing the type of the exception that was raised with the type that's expected.

**Comment by collenjones** [comment]
I have not generated any code, this is all through `tchannel.thrift.load`.

===Loading the IDL===
```
sys.modules[__name__] = thrift.load(
    path=...,
    service='servicename'
)
```

This is how I'm importing `InvalidArgument` in the testfile:
```
from servicename.thrift.servicename import InvalidArgument
```

===Exception Type===
```
tests/handlers/test_data_aggregation.py .> /home/uber/rubberstamp/rubberstamp/lib/thrift_handler_wrapper.py(25)wrapper()
     24             import ipdb; ipdb.set_trace()
---> 25             raise e
     26         except Exception as e:

ipdb> type(e)
<class 'thriftrw.compile.link.InvalidArgument'>
```

Really appreciate your help with this!

**Comment by abhinav** [comment]
Oh I see the problem:

            yield data_aggregation.tripCreatedOrChanged(request)

That's calling the handler function directly. The handler function has been decorated by TChannel to *always* return a Response object if the exception raised was an application exception (i.e., it was added under the `throws` section in the Thrift IDL). We have plans to make handler functions callable with request objects for tests (#144), but that's not yet supported.

If you do

    response = yield data_aggregation.tripCreatedOrChanged(request)

You'll see an object with class `DataAggregationService_userCreatedOrChanged_response` that has the `invalidArgument` field set to the exception.

**Comment by abhinav** [comment]
For the record, a workaround similar to the following can be used right now:

```python
def thriftrw_reraise(response):
    type_spec = response.body.__class__.type_spec
    for exc_spec in type_spec.exception_specs:
        exc = getattr(response.body, exc_spec.name)
        if exc:
            raise exc
    if type_spec.return_spec is not None:
        return response.body.success
```

It will re-raise the exception wrapped in the Response object. 

Meanwhile we expect to have a fix in 0.21.

**Comment by abhinav** [comment]
The fix for this was released in 0.21. See https://tchannel-python.readthedocs.org/en/latest/changelog.html#from-0-20-to-0-21 for how to upgrade.
